File: backend/app/schemas/v2/auto_allocate.py
"""Auto-allocate (Wave 41 v1.0 / auto-schedule) Pydantic schemas.

設計仕様書: ``docs/plans/auto-schedule-v1.md`` (v0.4) §10 データ書込仕様
実装計画書: ``docs/plans/auto-schedule-implementation-plan.md`` (v0.2) §4 Phase 2

エンドポイント:
  - POST /api/v1/schedule/auto-allocate
      request : ``AutoAllocateRequest``
      response: ``AutoAllocateResponse``
  - POST /api/v1/schedule/proposal/{proposal_batch_id}/apply
      response: ``ProposalApplyResponse``
  - POST /api/v1/schedule/proposal/{proposal_batch_id}/discard
      response: ``ProposalDiscardResponse``

実装方針:
  - auto_allocator.auto_allocate(...) の戻り値 dict をそのまま map できる
    形にフィールドを揃える.
  - mode は v1.0 では "mode_1" のみ実体実装。"mode_2" は service 層で
    NotImplementedError → API 層で 501 に変換 (本 schema 自体は許容).
"""

from __future__ import annotations

import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class AutoAllocateRequest(BaseModel):
    """``POST /api/v1/schedule/auto-allocate`` のリクエストボディ."""

    model_config = ConfigDict(extra="forbid")

    iso_year: int = Field(ge=2020, le=2100)
    iso_week: int = Field(ge=1, le=53)
    # H3 (W41 v1.0 cross-review): 異常な巨大リクエストを防ぐため最大 50 拠点に制限.
    office_ids: list[uuid.UUID] = Field(default_factory=list, max_length=50)
    mode: Literal["mode_1", "mode_2"] = "mode_1"


class KpiResponse(BaseModel):
    """``auto_allocate`` 戻り値の KPI dict と 1:1 対応するレスポンス型."""

    model_config = ConfigDict(extra="forbid")

    total_distance_km_before: float
    total_distance_km_after: float
    distance_reduction_pct: float
    staff_load_stddev_before: float
    staff_load_stddev_after: float
    # ハード制約 H1〜H8 の違反件数 (キーは "H1"…"H8").
    h_violations: dict[str, int]
    improvement_score: float


class ProposedCourseSummary(BaseModel):
    """1 提案コースのサマリ (auto_allocator proposals の API 公開形)."""

    model_config = ConfigDict(extra="forbid")

    course_id: uuid.UUID
    office_id: uuid.UUID
    weekday: int = Field(ge=0, le=6)
    code: str
    assigned_staff_id: uuid.UUID | None = None
    # M2: UI から staff の表示名を引きやすくするための補助フィールド.
    # auto_allocate endpoint は Staff を join して埋める. 不明な場合は None.
    assigned_staff_name: str | None = None
    visits_count: int = Field(ge=0)


class AutoAllocateResponse(BaseModel):
    """``POST /api/v1/schedule/auto-allocate`` のレスポンス."""

    model_config = ConfigDict(extra="forbid")

    proposal_batch_id: uuid.UUID
    proposals: list[ProposedCourseSummary] = Field(default_factory=list)
    kpi: KpiResponse
    warnings: list[str] = Field(default_factory=list)


class ProposalApplyResponse(BaseModel):
    """``POST /api/v1/schedule/proposal/{proposal_batch_id}/apply`` のレスポンス.

    提案バッチを採用し ``proposed → course_fixed → staff_assigned`` に遷移させた
    結果の件数を返す.
    """

    model_config = ConfigDict(extra="forbid")

    proposal_batch_id: uuid.UUID
    courses_committed: int = Field(ge=0)
    visits_committed: int = Field(ge=0)
    # C2: apply で再投入はしない. **既に auto_allocate 段階で INSERT 済**の
    # visit_staff_assignments の件数 (= 採用された visits に対応する行数) を返す.
    staff_assignments_committed: int = Field(ge=0)
    # M3: 採用された course_id のリスト (UI 側で後続 fetch を絞り込むための補助情報).
    course_ids: list[uuid.UUID] = Field(default_factory=list)


class ProposalDiscardResponse(BaseModel):
    """``POST /api/v1/schedule/proposal/{proposal_batch_id}/discard`` のレスポンス.

    提案バッチに紐付く未採用 (proposed) course + visit を物理削除した件数を返す.
    """

    model_config = ConfigDict(extra="forbid")

    proposal_batch_id: uuid.UUID
    courses_deleted: int = Field(ge=0)
    visits_deleted: int = Field(ge=0)


__all__ = [
    "AutoAllocateRequest",
    "AutoAllocateResponse",
    "KpiResponse",
    "ProposalApplyResponse",
    "ProposalDiscardResponse",
    "ProposedCourseSummary",
]

File: frontend/lib/schemas/v2/autoAllocate.ts
/**
 * Auto-allocate (Wave 41 v1.0) zod schemas.
 *
 * Backend `backend/app/schemas/v2/auto_allocate.py` の Pydantic schema と
 * **完全一致** させる。エンドポイント:
 *   - POST /api/v1/schedule/auto-allocate
 *   - POST /api/v1/schedule/proposal/{batch_id}/apply
 *   - POST /api/v1/schedule/proposal/{batch_id}/discard
 *
 * 設計仕様書: ``docs/plans/auto-schedule-v1.md`` v0.4 §10 データ書込仕様
 * 実装計画書: ``docs/plans/auto-schedule-implementation-plan.md`` v0.2 §4 Phase 2
 */
import { z } from 'zod';

// ---------------------------------------------------------------------------
// 算出モード — BE Literal['mode_1', 'mode_2'] と一致。
//   - mode_1: v1.0 で実装される実体ロジック
//   - mode_2: v1.0 では NotImplementedError → 501 (UI 上は選択肢として残す)
// ---------------------------------------------------------------------------
export const autoAllocateModeEnum = z.enum(['mode_1', 'mode_2']);
export type AutoAllocateMode = z.infer<typeof autoAllocateModeEnum>;

// ---------------------------------------------------------------------------
// Request — POST /api/v1/schedule/auto-allocate
// ---------------------------------------------------------------------------
export const autoAllocateRequestSchema = z.object({
  iso_year: z.number().int().min(2020).max(2100),
  iso_week: z.number().int().min(1).max(53),
  /** 空配列 = 全拠点。指定があれば対象拠点を絞る。 */
  office_ids: z.array(z.string().uuid()).default([]),
  mode: autoAllocateModeEnum.default('mode_1'),
});
export type AutoAllocateRequest = z.infer<typeof autoAllocateRequestSchema>;

// ---------------------------------------------------------------------------
// KPI — auto_allocate 戻り値の KPI dict と 1:1 対応。
// ---------------------------------------------------------------------------
export const kpiResponseSchema = z.object({
  total_distance_km_before: z.number(),
  total_distance_km_after: z.number(),
  distance_reduction_pct: z.number(),
  staff_load_stddev_before: z.number(),
  staff_load_stddev_after: z.number(),
  /** ハード制約 H1〜H8 の違反件数 (キーは "H1"…"H8")。 */
  h_violations: z.record(z.string(), z.number().int()),
  improvement_score: z.number(),
});
export type KpiResponse = z.infer<typeof kpiResponseSchema>;

// ---------------------------------------------------------------------------
// ProposedCourseSummary — 1 提案コースのサマリ。
// ---------------------------------------------------------------------------
export const proposedCourseSummarySchema = z.object({
  course_id: z.string().uuid(),
  office_id: z.string().uuid(),
  /** 曜日 (0=月..6=日) */
  weekday: z.number().int().min(0).max(6),
  code: z.string(),
  assigned_staff_id: z.string().uuid().nullable().optional(),
  /** UI 表示用の staff 名 (BE 側で join して埋める。未解決なら null)。 */
  assigned_staff_name: z.string().nullable().optional(),
  visits_count: z.number().int().nonnegative(),
});
export type ProposedCourseSummary = z.infer<typeof proposedCourseSummarySchema>;

// ---------------------------------------------------------------------------
// Response — POST /api/v1/schedule/auto-allocate
// ---------------------------------------------------------------------------
export const autoAllocateResponseSchema = z.object({
  proposal_batch_id: z.string().uuid(),
  proposals: z.array(proposedCourseSummarySchema).default([]),
  kpi: kpiResponseSchema,
  warnings: z.array(z.string()).default([]),
});
export type AutoAllocateResponse = z.infer<typeof autoAllocateResponseSchema>;

// ---------------------------------------------------------------------------
// Response — POST /api/v1/schedule/proposal/{batch_id}/apply
// ---------------------------------------------------------------------------
export const proposalApplyResponseSchema = z.object({
  proposal_batch_id: z.string().uuid(),
  courses_committed: z.number().int().nonnegative(),
  visits_committed: z.number().int().nonnegative(),
  staff_assignments_committed: z.number().int().nonnegative(),
  course_ids: z.array(z.string().uuid()).default([]),
});
export type ProposalApplyResponse = z.infer<typeof proposalApplyResponseSchema>;

// ---------------------------------------------------------------------------
// Response — POST /api/v1/schedule/proposal/{batch_id}/discard
// ---------------------------------------------------------------------------
export const proposalDiscardResponseSchema = z.object({
  proposal_batch_id: z.string().uuid(),
  courses_deleted: z.number().int().nonnegative(),
  visits_deleted: z.number().int().nonnegative(),
});
export type ProposalDiscardResponse = z.infer<typeof proposalDiscardResponseSchema>;
